Classic themes: prevent access to parts of the Site Editor

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->

This PR removes some content from specific screens in the Site Editor when a classic theme is active.
Note: This is an edge case since it means that the user is manually navigating to a screen that there is no direct link to in the default WordPress UI.

Closes https://github.com/WordPress/gutenberg/issues/68950
Alternative to https://github.com/WordPress/gutenberg/pull/68959

<!-- In a few words, what is the PR actually doing? -->

## Why?
There are Site Editor features that doesn't work with classic themes yet, such as creating and editing block templates
or previewing the navigation menu.

## How?
The PR adds checks for the theme type, and then hides the sidebar and content if the theme is not a block theme.

## Testing Instructions
Activate a block theme and navigate to different Site Editor pages to check that there are no regressions.

Activate a classic theme.
These screens also needs to be tested on narrow screen widths.

Navigate to:
`wp-admin/site-editor.php?p=%2Ftemplate`
`wp-admin/site-editor.php?p=%2Fnavigation`
`wp-admin/site-editor.php?p=%2Fstyles`
`wp-admin/site-editor.php?p=%2Fpage`

singe page:
`wp-admin/site-editor.php?p=%2Fpage&postId= YOUR ID HERE`

single navigation menu:
`wp-admin/site-editor.php?p=%2Fwp_navigation%2FYOUR ID HERE`

There should be a message that says: "The theme you are currently using is not compatible with the Site Editor."

## Screenshots or screencast <!-- if applicable -->

Updated screenshot February 24: The navigation screen showing the notice instead of the sidebar content.
<img width="1503" alt="Site Editor navigation screen" src="https://github.com/user-attachments/assets/3906ac20-7f77-490f-a482-cc345cf92e1e" />

carolinan committed:
Classic themes: prevent access to parts of the Site Editor

carolinan commented:
For the Pages page, I wonder if access to the data views could be allowed but hide the option to create new templates.

I mean this option, when you are editing a page and click on the template name:
<img width="293" alt="the create template option in the page settings sidebar" src="https://github.com/user-attachments/assets/410984ba-7618-400a-bdbf-8ddb758ecc72" />

I am not sure what else may be broken for classic themes.
Again, we are talking about users that are manually navigating to Pages by changing the path.
Not about now making the pages screen available to everyone. I will leave that to another discussion.

carolinan committed:
Merge branch 'trunk' into fix/classic-site-editor-access

carolinan committed:
Add spacing around the notice.

carolinan committed:
Limit access to the global styles options

carolinan committed:
global styles: Make sure the notice is visible on narrow screens

carolinan committed:
Limit access to the navigation screen

carolinan commented:
I am not found of this solution.
Maybe instead the "App" could update the path and direct the user to the Not Found screen? I have not tried it.

carolinan commented:
I could use some help with figuring out what needs to be tested with hybrid themes.

t-hamano commented:
> Again, we are talking about users that are manually navigating to Pages by changing the path.
> Not about now making the pages screen available to everyone by showing the menu item. I will leave that to another discussion.

I think that for screens that are accessible via URL but were previously prohibited from being accessed from classic/hybrid themes, only an error message should be displayed.

> I could use some help with figuring out what needs to be tested with hybrid themes.

Does this mean which screens we should test? If so, [this route list](https://github.com/WordPress/gutenberg/tree/8c0db8e5b1dda64c317bbcec86896c675b49b0f9/packages/edit-site/src/components/site-editor-routes) should help.

carolinan commented:
I have not worked with hybrid themes for a long time, I am unsure of what features are intended to be available:
with theme.json
with the block template theme support

carolinan committed:
Prevent access to the pages screen

carolinan committed:
Pages: Switch the returns so that the message shows on narrow screens.

carolinan commented:
I am working on hiding the preview of the single pages, since you can click on the preview and edit the page in the Site Editor. But when I log the post type value on the single page screen, it actually uses `wp_template`, not page. This was unexpected.

carolinan committed:
Prevent access to the single navigation screen

carolinan commented:
To test hybrid themes, I added `add_theme_support( 'block-templates' );` to Twenty Ten, and added a few HTML files in the templates folder.
The way hybrid themes edit templates is by using the **block editor**, not the Site Editor. The changes here that prevents access to the Site Editor templates screen should have no negative effect on this feature, and I don't believe further testing is needed.

carolinan commented:
Still, I am not so found of this code repetition, there may be better ways to solve this.

carolinan commented:
A lot of time spend on this issue was used to identify which file contained the code for which screen. I wonder if we could make this easier. What would be a suitable way to document this?

t-hamano commented on packages/edit-site/src/components/sidebar-navigation-screen-templates-browse/index.js:
```diff
@@ -10,14 +12,24 @@ import SidebarNavigationScreen from '../sidebar-navigation-screen';
+							'Create new templates, or reset any customizations made to the templates supplied by your theme.'
+					  )
+					: null
+			}
```
```suggestion
			description={
				isBlockBasedTheme &&
				__(
					'Create new templates, or reset any customizations made to the templates supplied by your theme.'
				)
			}
```

Regarding the `content` prop, I think it can be simplified by using `&&`.

t-hamano reviewed (commented):
I think that instead of adding conditional statements to the `Editor` component or `PageTemplates` component themselves, we should add similar processing to [the routes](https://github.com/WordPress/gutenberg/tree/6d8753f56f7ba0b48427ebd09987eaa16efecf76/packages/edit-site/src/components/site-editor-routes) itself. This should prevent unnecessary hooks from being executed.

Taking the Navigation route as an example, we could probably implement it like this:

<details><summary>Example</summary>

```jsx
function NavigationContent() {
	// In the classic theme, display error text
	if ( ! isBlockBasedTheme ) {
		return __( 'The theme you are currently using is not compatible with the Site Editor.' );
	}
	// In the block theme, we don't want to show any content
	return null;
}

function NavigationPreview() {
	// In the classic theme, we don't want to show any preview
	if ( ! isBlockBasedTheme ) {
		return null;
	}
	// In the block theme, render the resizable editor preview
	return <Editor />;
}

export const navigationRoute = {
	name: 'navigation',
	path: '/navigation',
	areas: {
		content: <NavigationContent />,
		preview: <NavigationPreview />,
	},
};
```

</details> 

Unfortunately, this approach doesn't work: the `areas.content` or `areas.preview` wrappers will always be rendered, even if the component returns null:

![navigation](https://github.com/user-attachments/assets/cbdb2e02-6c45-4944-a13a-5b12b50f61d7)

@youknowriad @tyxla @jsnajdr 

We want to conditionally render `areas.content` and `areas.preview` depending on whether the theme is a block theme or not. For example, we may want to render only `areas.content` in the classic theme and only `areas.preview` in the block theme.

However, we cannot completely hide it because the wrapper element is always rendered just by defining the `content` or `preview` field.

If you know any good approaches, please let us know 🙏

carolinan commented:
I was thinking of adding the theme type check here and updating the path to "not found".
https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/app/index.js#L37

t-hamano commented:
> I was thinking of adding the theme type check here and updating the path to "not found".
> I have **not** tried it yet.
> https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/app/index.js#L37

I'm a bit unsure about this, but it looks like a good approach to me. I tested it briefly, and if I returned a path that didn't exist, I was able to navigate to the not found screen:

<details><summary>Example</summary>

Note: The `isAllowedPath` function needs to be created.

```jsx
export default function App() {
	// ...
	const { routes, isBlockBasedTheme } = useSelect( ( select ) => {
		return {
			routes: unlock( select( editSiteStore ) ).getRoutes(),
			isBlockBasedTheme:
				select( coreStore ).getCurrentTheme()?.is_block_theme,
		};
	}, [] );
	const beforeNavigate = useCallback( ( { path, query } ) => {
		if ( ! isBlockBasedTheme && ! isAllowedPath( '/stylebook' ) ) {
			return { path: '/404' };
		}
		// ...
	}, [] );

	// ...
}
```

</details>

carolinan committed:
Merge branch 'trunk' into fix/classic-site-editor-access

carolinan commented:
It doesn't work since the user is manually entering the URL, and not "navigating" . I am explaining that really bad :)

carolinan committed:
Simplify conditions

carolinan committed:
Remove the limitation that was added to the editor

This limitation also removed the preview on the start screen.

carolinan committed:
Revert changes to EditSiteEditor

carolinan committed:
Revert changes to EditSiteEditor

carolinan committed:
Use a warning notice for the compatibility warning

carolinan committed:
Try to update the templatesRoute instead of PageTemplates

carolinan committed:
Try to update pagesRoute instead of PostList

carolinan committed:
Remove the unused CSS

carolinan committed:
Update NavigationItemRoute to hide the preview of the single navigation item

carolinan committed:
Hide the page preview

carolinan committed:
Merge branch 'trunk' into fix/classic-site-editor-access

t-hamano commented:
> We want to conditionally render `areas.content` and `areas.preview` depending on whether the theme is a block theme or not. For example, we may want to render only `areas.content` in the classic theme and only `areas.preview` in the block theme.
> 
> However, we cannot completely hide it because the wrapper element is always rendered just by defining the `content` or `preview` field.

I've been thinking about ways to solve this problem, and the idea I came up with is to add the following logic to [the `Layout` component](https://github.com/WordPress/gutenberg/blob/759ef8721a8bd329aedb5eb679de6801d638b313/packages/edit-site/src/components/layout/index.js) itself.

This should allow you to completely hide both the content and the preview when you access the Pages screen in the classic theme, for example. However, I'm not sure if this is the right approach 🤷

<details><summary>Example</summary>

Note that I have simplified the code.

```jsx
function Layout() {
	const { query, name: routeKey, areas, widths } = useLocation();

	const isBlockBasedTheme = useSelect(
		( select ) => select( coreStore ).getCurrentTheme()?.is_block_theme,
		[]
	);

	const screensForClassicTheme = [ 'patterns', 'styles' ];
	const shouldRenderContentArea =
		areas.content &&
		( isBlockBasedTheme || ! screensForClassicTheme.includes( routeKey ) );
	const shouldRenderPreviewArea =
		areas.preview &&
		( isBlockBasedTheme || ! screensForClassicTheme.includes( routeKey ) );

	return (
		<>
			<div className="edit-site-layout">
				<div className="edit-site-layout__content">
					{ ! isMobileViewport &&
						shouldRenderContentArea &&
						canvas !== 'edit' && (
							<div className="edit-site-layout__area"></div>
						) }
					{ ! isMobileViewport && shouldRenderPreviewArea && (
						<div className="edit-site-layout__canvas-container"></div>
					) }
				</div>
			</div>
		</>
	);
}
```

</details> 


Additional ping: @stokesman  I believe you've been working on improving the site editor screen lately 🙏

carolinan commented:
If it is hidden, there wont be a message showing that the use does not have access 🤔

t-hamano commented:
> If it is hidden, there wont be a message showing that the user does not have access 🤔

Sorry for the poor explanation, I was thinking of moving the error notification to the sidebar:

![image](https://github.com/user-attachments/assets/d4819040-132b-4f05-a9c0-4ac5b4f753db)

carolinan commented:
It doesn't work because its not visible on the narrow screens.

t-hamano commented:
Can we display the same message by using [`areas.mobile`](https://github.com/WordPress/gutenberg/pull/69005/files#diff-ef4afd1b4d7fede276380b507e003015411d52c3d8394c8cdb76bbfdc8d36118L44)?

t-hamano commented:
Would it be ok to try a different approach in a different PR? I think I'll have it ready by tomorrow at the latest.

carolinan commented:
Yes I think it would be easiest to test it on a clean branch based on trunk.

t-hamano commented:
I came up with the idea that a more ideal solution would be to extend the router package itself: #69299

carolinan committed:
Merge branch 'trunk' into fix/classic-site-editor-access

t-hamano commented:
Thanks for working on this issue. I've submitted #69473, which is based on the new approach.
